### PR TITLE
voicemails: add greetings API endpoints

### PR DESCRIPTION
**Pull request by sileht**

### CODE REVIEW

File: wazo_calld/plugins/voicemails/exceptions.py [hunk 1]
```diff
@@ -75,3 +75,29 @@ def __init__(self):
             message='Invalid voicemail message format',
             error_id='invalid-voicemail-message-format',
         )
+
+
+class InvalidVoicemailGreeting(APIException):
+
+    def __init__(self, greeting):
+        super().__init__(
+            status_code=400,
```
Comment (fblackburn1) [inline wazo_calld/plugins/voicemails/exceptions.py]: 404

File: wazo_calld/plugins/voicemails/api.yml [hunk 2]
```diff
@@ -376,10 +358,7 @@ paths:
           $ref: '#/responses/AnotherServiceUnavailable'
     head:
       summary: Check if a custom greeting exist
-      description: '**Required ACL:** `calld.users.me.voicemails.greetings.{greeting}.read`
-
-
-        **Not implemented**'
+      description: '**Required ACL:** `calld.users.me.voicemails.greetings.{greeting}.read`'
```
Comment (fblackburn1) [inline wazo_calld/plugins/voicemails/api.yml]: The HEAD doesn't seem implemented?

File: wazo_calld/plugins/voicemails/resources.py [hunk 3]
```diff
@@ -254,3 +264,107 @@ def _validate_message_id(message_id):
...
+    def post(self, greeting):
+        voicemail_id = _get_voicemail_id_from_request(self._auth_client, self._voicemails_service)
+        greeting = _validate_greeting(greeting)
+        return self._copy(voicemail_id, greeting)
+
+
+def _validate_greeting(greeting):
+    if greeting in VALID_GREETINGS:
+        return greeting
+    raise InvalidVoicemailGreeting(greeting)
```
Comment (fblackburn1) [inline wazo_calld/plugins/voicemails/resources.py]: Use marshmallow and try-except to reraise the good exception OR raise the exception from a custom marshmallow validator (with `@validate(<field>)` decorator).
If it not necessary, do not make field validation outside schema

File: integration_tests/suite/test_voicemail.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+from .helpers.base import RealAsteriskIntegrationTest
+from .helpers.confd import (
+    MockUser,
+    MockVoicemail,
+)
+from .helpers.constants import ASSET_ROOT, VALID_TENANT
+from .helpers.hamcrest_ import HamcrestARIChannel
+
+VALID_GREETINGS = ('busy', 'unavailable', 'name')
+WAVE_FILE = os.path.join(ASSET_ROOT, 'bugs_29.wav')
```
[the remark this answers is not in the document]
Comment (sileht) [reply integration_tests/suite/test_voicemail.py]: Sample 26 of Bugs bunny free wave sound :smiley:

File: integration_tests/suite/test_voicemail.py [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+        self.calld_client.set_token(self._user_token)
+
+    def test_voicemail_get_greeting_invalid_voicemail(self):
+        assert_that(
+            calling(self.calld_client.voicemails.get_voicemail_greeting).with_args(
+                "not-exists", "busy"
+            ),
+            raises(requests.HTTPError).matching(has_property(
+                'response', has_properties(
+                    status_code=400,
```
[the remark this answers is not in the document]
Comment (sileht) [reply integration_tests/suite/test_voicemail.py]: Oh yeah :)

File: wazo_calld/plugins/voicemails/api.yml [hunk 6]
```diff
@@ -587,4 +557,4 @@ parameters:
     enum:
     - unavailable
     - busy
-    - temp
+    - name
```
[the remark this answers is not in the document]
Comment (sileht) [reply wazo_calld/plugins/voicemails/api.yml]: No idea, the @nballas-wazo change use `name` and not `temp`. So I changed it here.

File: wazo_calld/plugins/voicemails/resources.py (hunk 3 above)
Comment (sileht) [reply wazo_calld/plugins/voicemails/resources.py]: I have copying the scheme of the rest of the file, and wondering why marshmallow was not used. I will fix all of them.

File: wazo_calld/plugins/voicemails/api.yml (hunk 2 above)
Comment (sileht) [reply wazo_calld/plugins/voicemails/api.yml]: I wonder if that make sense to do it, as asterisk doesn't provide HEAD neither.

@fblackburn1 @pc-m @nballas-wazo What do you think, we implement it in Asterisk or we drop it here ?

File: wazo_calld/plugins/voicemails/exceptions.py (hunk 1 above)
Comment (sileht) [reply wazo_calld/plugins/voicemails/exceptions.py]: fixed

File: wazo_calld/plugins/voicemails/api.yml [hunk 7]
```diff
@@ -145,10 +142,7 @@ paths:
           $ref: '#/responses/AnotherServiceUnavailable'
     head:
       summary: Check if a custom greeting exist
-      description: '**Required ACL:** `calld.voicemails.{voicemail_id}.greetings.{greeting}.read`
-
-
-        **Not implemented**'
+      description: '**Required ACL:** `calld.voicemails.{voicemail_id}.greetings.{greeting}.read`'
```
[the remark this answers is not in the document]
Comment (sileht) [reply wazo_calld/plugins/voicemails/api.yml]: See comment bellow.

File: integration_tests/suite/test_voicemail.py [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+        )
+
+    def test_voicemail_get_greeting_invalid_greeting(self):
+        assert_that(
+            calling(self.calld_client.voicemails.get_voicemail_greeting).with_args(
+                self._voicemail_id, "not-exists"
+            ),
+            raises(requests.HTTPError).matching(has_property(
+                'response', has_properties(
+                    status_code=400,
```
[the remark this answers is not in the document]
Comment (sileht) [reply integration_tests/suite/test_voicemail.py]: fixed.

File: integration_tests/suite/helpers/confd.py [hunk 9]
```diff
@@ -73,6 +73,22 @@ def set_moh(self, *mock_mohs):
...
+        body = {
+            'response': 'user_voicemails',
+            'content': dict([
+                (user_uuid, {'user_id': user_uuid,
+                             'voicemail_id': voicemail.id()})
+                for voicemail in mock_voicemails
+                for user_uuid in voicemail.user_uuids
+            ])
+        }
+        requests.post(url, json=body, verify=False)
```
[the remark this answers is not in the document]
Comment (sileht) [reply integration_tests/suite/helpers/confd.py]: `voicemails` are untested, and already uses `user_voicemails`, I haven't added that, just mocked it to be able to test it.

I will try to remove it, that looks small at first glance.

[the remark this answers is not in the document]
Comment (sileht) [reply integration_tests/suite/helpers/confd.py]: I have checked and the issue looks to be in confd client not calld it self.

[the remark this answers is not in the document]
Comment (sileht) [reply integration_tests/suite/helpers/confd.py]: https://github.com/wazo-pbx/wazo-calld/blob/master/wazo_calld/helpers/confd.py#L134-L142

[the remark this answers is not in the document]
Comment (sileht) [reply integration_tests/suite/helpers/confd.py]: https://github.com/wazo-pbx/wazo-confd-client/blob/master/wazo_confd_client/relations.py#L43

[the remark this answers is not in the document]
Comment (sileht) [reply integration_tests/suite/helpers/confd.py]: I added a commit that remove its usage.

File: integration_tests/suite/test_voicemail.py [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+        )
+
+    def test_voicemail_get_greeting_invalid_greeting_from_user(self):
+        assert_that(
+            calling(self.calld_client.voicemails.get_voicemail_greeting_from_user).with_args(
+                "not-exists"
+            ),
+            raises(requests.HTTPError).matching(has_property(
+                'response', has_properties(
+                    status_code=400,
```
[the remark this answers is not in the document]
Comment (sileht) [reply integration_tests/suite/test_voicemail.py]: fixed

**Comment by sileht** [comment]
> Should we catch all ari exception to re-raise well formatted 500 error?

That already done: https://github.com/wazo-pbx/wazo-calld/blob/master/wazo_calld/http.py#L26-L35

File: integration_tests/suite/helpers/confd.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (fblackburn1) [reply integration_tests/suite/helpers/confd.py]: > `voicemails` are untested, and already uses `user_voicemails`, I haven't added that, just mocked it to be able to test it.

Oh, as I said, I didn't validate my comments :trollface:

File: wazo_calld/helpers/confd.py [hunk 11]
```diff
@@ -133,7 +133,9 @@ def from_id(cls, conference_id, confd_client):
 
 def get_user_voicemail(user_uuid, confd_client):
     try:
-        return confd_client.users.relations(user_uuid).get_voicemail()
+        return confd_client.voicemails.list(user_uuid=user_uuid)['items'][0]
```
Comment (fblackburn1) [inline wazo_calld/helpers/confd.py]: I've not tested, but I think is not working: `user_uuid=user_uuid`
Why not using `confd_client.users.get(user_uuid)['voicemail']` ?

File: wazo_calld/plugins/voicemails/api.yml (hunk 2 above)
Comment (fblackburn1) [reply wazo_calld/plugins/voicemails/api.yml]: for record, solution is to drop it until we need it

File: wazo_calld/plugins/voicemails/resources.py [hunk 12]
```diff
@@ -49,10 +56,15 @@ class VoicemailMessageUpdateSchema(Schema):
     folder_id = fields.Integer(required=True)
 
 
+class VoicemailGreetingCopy(Schema):
```
Comment (fblackburn1) [inline wazo_calld/plugins/voicemails/resources.py]: The convention is to add `Schema` suffix for schema class

File: wazo_calld/plugins/voicemails/resources.py (hunk 3 above)
Comment (fblackburn1) [reply wazo_calld/plugins/voicemails/resources.py]: Oh you right, it is the `VoicemailGreetingCopy` without validation for greeting that confused me. But you have fixed it :+1:

File: wazo_calld/plugins/voicemails/resources.py (hunk 12 above)
Comment (sileht) [reply wazo_calld/plugins/voicemails/resources.py]: done

File: wazo_calld/helpers/confd.py (hunk 11 above)
Comment (sileht) [reply wazo_calld/helpers/confd.py]: Indeed that looks better :)

Comment (fblackburn1) [reply wazo_calld/helpers/confd.py]: And I also confirm, that you cannot filter by `user_uuid` on the voicemail resource :)

**Comment by fblackburn1** [comment]
I got a `500 {"message": "Internal Server Error"}` when using swagger-ui. I expect a more readable message

Also can you open a ticket to fetch default asterisk greeting when user doesn't set greeting?

File: wazo_calld/plugins/voicemails/api.yml [hunk 13]
```diff
@@ -320,6 +320,25 @@ paths:
...
+      - voicemails
+      responses:
+        '204':
+          description: The greeting was created successfully
+        '400':
+          $ref: '#/responses/InvalidRequest'
+        '404':
+          $ref: '#/responses/NotFoundError'
+        '503':
+          $ref: '#/responses/AnotherServiceUnavailable'
```
Comment (fblackburn1) [inline wazo_calld/plugins/voicemails/api.yml]: documentation missing for `/voicemails/{voicemail_id}/greetings/{greeting}`

File: wazo_calld/plugins/voicemails/resources.py [hunk 14]
```diff
@@ -273,6 +273,13 @@ class _BaseVoicemailGreetingResource(AuthResource):
     def __init__(self, voicemails_service):
         self._voicemails_service = voicemails_service
 
+    def _post(self, voicemail_id, greeting):
+        # FIXME(sileht):
+        # data = voicemail_greeting_update_schema.load(request.get_json(force=True)).data
```
Comment (fblackburn1) [inline wazo_calld/plugins/voicemails/resources.py]: ??

File: wazo_calld/plugins/voicemails/services.py [hunk 15]
```diff
@@ -81,17 +84,40 @@ def get_greeting(self, voicemail_id, greeting):
...
+        }
+        try:
+            self._ari.wazo.createVoicemailGreeting(
+                context=vm_conf['context'],
+                voicemail=vm_conf['name'],
+                greeting=greeting,
+                body=body
+            )
+        except ARIHTTPError as e:
+            # FIXME(sileht): Should be 409 or 400
```
Comment (fblackburn1) [inline wazo_calld/plugins/voicemails/services.py]: Why is not the case? not merged or it's a bug in the asterisk patch?

File: wazo_calld/plugins/voicemails/exceptions.py [hunk 16]
```diff
@@ -101,3 +101,15 @@ def __init__(self, greeting):
...
+class VoicemailGreetingAlreadyExists(APIException):
+    def __init__(self, greeting):
+        super().__init__(
+            status_code=409,
+            message='Voicemail greeting already exists',
+            error_id='voicemail-greeting-already-exists',
+            details={
+                'greeting': greeting,
+            }
+        )
```
Comment (fblackburn1) [inline wazo_calld/plugins/voicemails/exceptions.py]: I think we should also set a key `"voicemail_id": voicemail_id` in `details, but we can wait until we need it

File: wazo_calld/plugins/voicemails/services.py (hunk 15 above)
Comment (sileht) [reply wazo_calld/plugins/voicemails/services.py]: done

File: wazo_calld/plugins/voicemails/api.yml (hunk 13 above)
Comment (sileht) [reply wazo_calld/plugins/voicemails/api.yml]: done

File: wazo_calld/plugins/voicemails/resources.py (hunk 14 above)
Comment (sileht) [reply wazo_calld/plugins/voicemails/resources.py]: removed

**Pull request merged**